pkg/experiment/metastore/index/index.go

```diff
@@ -0,0 +1,511 @@
+				}
+				sh.tenants[t] = te
+			}
+		}
+	}
+
+	return p, nil
+}
+
+// CreatePartitionKey creates a partition key for a block. It is meant to be used for newly inserted blocks, as it relies
+// on the index's currently configured partition duration to create the key.
+//
+// Note: Using this for existing blocks following a partition duration change can produce the wrong key. Callers should
+// verify that the returned partition actually contains the block.
+func (i *Index) CreatePartitionKey(blockId string) PartitionKey {
+	t := ulid.Time(ulid.MustParse(blockId).Time()).UTC()
+
+	var b strings.Builder
+	b.Grow(16)
+
+	year, month, day := t.Date()
+	b.WriteString(fmt.Sprintf("%04d%02d%02d", year, month, day))
+
+	partitionDuration := i.Config.PartitionDuration
+	if partitionDuration < 24*time.Hour {
+		hour := (t.Hour() / int(partitionDuration.Hours())) * int(partitionDuration.Hours())
+		b.WriteString(fmt.Sprintf("T%02d", hour))
+	}
+
+	mDuration := model.Duration(partitionDuration)
+	b.WriteString(".")
+	b.WriteString(mDuration.String())
+
+	return PartitionKey(b.String())
+}
+
+// FindPartitionMeta retrieves the partition meta for the given key.
+func (i *Index) FindPartitionMeta(key PartitionKey) *PartitionMeta {
+	loaded, ok := i.loadedPartitions[key]
+	if ok {
+		return loaded.meta
+	}
+	for _, p := range i.allPartitions {
+		if p.Key == key {
+			return p
+		}
+	}
+	return nil
+}
+
+// InsertBlock is the primary way for adding blocks to the index.
+func (i *Index) InsertBlock(b *metastorev1.BlockMeta) error {
+	i.partitionMu.Lock()
+	defer i.partitionMu.Unlock()
+
+	return i.insertBlock(b)
+}
+
+// insertBlock is the underlying implementation for inserting blocks. It is the caller's responsibility to enforce safe
+// concurrent access. The method will create a new partition if needed.
+func (i *Index) insertBlock(b *metastorev1.BlockMeta) error {
+	meta, err := i.GetOrCreatePartitionMeta(b)
+	if err != nil {
+		return err
+	}
+
+	p := i.getOrCreatePartition(meta)
+
+	p.shardsMu.Lock()
+	defer p.shardsMu.Unlock()
+
+	s, ok := p.shards[b.Shard]
+	if !ok {
+		s = &indexShard{
+			tenants: make(map[string]*indexTenant),
+		}
+		p.shards[b.Shard] = s
+	}
+
+	s.tenantsMu.Lock()
+	defer s.tenantsMu.Unlock()
+
+	ten, ok := s.tenants[b.TenantId]
+	if !ok {
+		ten = &indexTenant{
+			blocks: make(map[string]*metastorev1.BlockMeta),
+		}
+		s.tenants[b.TenantId] = ten
+	}
+
+	ten.blocksMu.Lock()
+	defer ten.blocksMu.Unlock()
+
+	ten.blocks[b.Id] = b
+	return nil
+}
+
+// GetOrCreatePartitionMeta makes the mapping between blocks and partitions. It may assign the block to an existing
+// partition or create a new partition altogether. Meant to be used only in the context of new blocks.
+func (i *Index) GetOrCreatePartitionMeta(b *metastorev1.BlockMeta) (*PartitionMeta, error) {
+	key := i.CreatePartitionKey(b.Id)
+	meta := i.FindPartitionMeta(key)
+
+	if meta == nil {
+		ts, duration, err := key.Parse()
+		if err != nil {
+			return nil, err
+		}
```

Comment:
`CreatePartitionKey` should not produce malformed keys. Block ID should be validated before any attempt to access the index